base_layer/core/src/transaction_protocol_manager.rs

```diff
@@ -0,0 +1,1046 @@
+struct TransactionProtocolStateData {
+    amount: Option<u64>,
+    lock_height: Option<u64>,
+    fee: Option<u64>,
+    inputs: Vec<TransactionInput>,
+    outputs: Vec<TransactionOutput>,
+    offset: Option<BlindingFactor>,
+    sender_excess_blinding_factor: Option<BlindingFactor>,
+    sender_excess: Option<PublicKey>,
+    sender_private_nonce: Option<SecretKey>,
+    sender_public_nonce: Option<PublicKey>,
+    receiver_output_blinding_factor: Option<BlindingFactor>,
+    receiver_output_public_key: Option<PublicKey>,
+    receiver_output: Option<TransactionOutput>,
+    receiver_private_nonce: Option<SecretKey>,
+    receiver_public_nonce: Option<PublicKey>,
+    receiver_partial_signature: Option<Signature>,
+    sender_partial_signature: Option<Signature>,
+    final_signature: Option<Signature>,
+}
+
+/// This is the message containing the public data that the Sender will send to the Receiver
+pub struct SenderPublicData {
+    pub fee: u64,
+    pub lock_height: u64,
+    pub amount: u64,
+    pub sender_excess: PublicKey,
+    pub sender_public_nonce: PublicKey,
+    pub offset: BlindingFactor,
+}
+
+/// This is the message containing the public data that the Receiver will send back to the Sender
+#[derive(Clone)]
+pub struct ReceiverPublicData {
+    pub receiver_output: TransactionOutput,
+    pub receiver_output_public_key: PublicKey,
+    pub receiver_public_nonce: PublicKey,
+    pub receiver_partial_signature: Signature,
+}
+
+/// In this state the Sender will start the protocol by supplying all of the data required from them
+/// Once all the data is supplied the final step is for the Sender to select their offset and the state
+/// machine will transition to the next state.
+/// The Receiver will also start in this state and will accept the public data from the Sender in order to
+/// transition to the next state.
+#[derive(Clone, Debug)]
+struct SenderSetup {
+    state_data: TransactionProtocolStateData,
+}
+
+impl SenderSetup {
+    fn new() -> SenderSetup {
+        let state_data = TransactionProtocolStateData {
+            amount: None,
+            lock_height: None,
+            fee: None,
+            inputs: Vec::new(),
+            outputs: Vec::new(),
+            offset: None,
+            sender_excess_blinding_factor: None,
+            sender_excess: None,
+            sender_private_nonce: None,
+            sender_public_nonce: None,
+            receiver_output_blinding_factor: None,
+            receiver_output_public_key: None,
+            receiver_output: None,
+            receiver_private_nonce: None,
+            receiver_public_nonce: None,
+            receiver_partial_signature: None,
+            sender_partial_signature: None,
+            final_signature: None,
+        };
+
+        SenderSetup { state_data }
+    }
+
+    /// This method accepts data from a Sender and transitions into the next state
+    /// TODO Check that this is the Receiver and has no Sender state set?
```

Comment:
Have you thought about having 2 different STNs? One for `Initiator` / Sender and one for `Receiver`? Then you don't have to do any checks here (and in the other places where you're having different transitions depending on the type of actor).

I'm also thinking forward to generalising this with 1 initiator and _n_ receivers; it might be easier to implement if the actors are separated.

Reply:
I did consider that but I also wanted to support this process being performed by a single actor (e.g. a wallet building a coin join) and it felt like a very subtle extension to allow for both use cases in a single state machine. However, I do agree with you that for future extension to ,multiple parties encoding this as two state-machines is a more modular design.